File: backend/src/services/llmService.js
const axios = require('axios');
const logger = require('../utils/logger');
const { Config, News } = require('../models');
const { decrypt } = require('../utils/encryption');

class LLMService {
  constructor() {
    this.apiUrl = null;
    this.apiKey = null;
    this.model = null;
  }

  // 初始化配置
  async initialize() {
    try {
      this.apiUrl = await this.getConfigValue('llm_api_url');
      this.apiKey = await this.getConfigValue('llm_api_key');
      this.model = await this.getConfigValue('llm_model');
      
      if (!this.apiUrl || !this.apiKey) {
        throw new Error('LLM API配置不完整');
      }
    } catch (error) {
      logger.error('LLM服务初始化失败:', error);
      throw error;
    }
  }

  // 获取配置值
  async getConfigValue(key, defaultValue = null) {
    try {
      const config = await Config.findOne({ where: { key } });
      if (!config) return defaultValue;
      
      let value = config.value;
      if (config.isEncrypted) {
        value = decrypt(value);
      }
      
      return value;
    } catch (error) {
      logger.error(`获取配置失败: ${key}`, error);
      return defaultValue;
    }
  }

  // 获取代理配置
  async getProxyConfig() {
    try {
      const httpProxyEnabled = await this.getConfigValue('http_proxy_enabled', false);
      const httpProxyUrl = await this.getConfigValue('http_proxy_url', '');
      
      if (httpProxyEnabled && httpProxyUrl) {
        return {
          host: new URL(httpProxyUrl).hostname,
          port: new URL(httpProxyUrl).port || 80,
          protocol: new URL(httpProxyUrl).protocol.replace(':', '')
        };
      }
      
      return null;
    } catch (error) {
      logger.error('获取代理配置失败:', error);
      return null;
    }
  }

  // 生成对话
  async generateDialogue(dialogueParams) {
    try {
      await this.initialize();
      
      const {
        dialogueType = 'interview',
        character1 = '主持人',
        character2 = '嘉宾',
        rounds = 8,
        newsCount = 5
      } = dialogueParams;

      // 获取最新新闻
      const news = await this.getLatestNews(newsCount);
      if (news.length === 0) {
        throw new Error('没有可用的新闻数据');
      }

      // 构建新闻摘要
      const newsSummary = this.buildNewsSummary(news);
      
      // 构建对话提示
      const prompt = this.buildDialoguePrompt(dialogueType, character1, character2, newsSummary, rounds);
      
      // 调用LLM API
      const response = await this.callLLMAPI(prompt);
      
      // 解析对话内容
      const dialogue = this.parseDialogueResponse(response, character1, character2);
      
      return {
        title: `${character1}与${character2}的${this.getDialogueTypeName(dialogueType)}`,
        content: dialogue,
        dialogueType,
        character1,
        character2,
        rounds,
        newsCount,
        news: news.map(n => ({
          id: n.id,
          title: n.title,
          source: n.sourceName
        }))
      };
    } catch (error) {
      logger.error('生成对话失败:', error);
      throw error;
    }
  }

  // 获取最新新闻
  async getLatestNews(count) {
    return await News.findAll({
      where: {
        isIgnored: false
      },
      order: [['publishedAt', 'DESC']],
      limit: count,
      attributes: ['id', 'title', 'summary', 'content', 'sourceName', 'category', 'publishedAt']
    });
  }

  // 构建新闻摘要
  buildNewsSummary(news) {
    return news.map((item, index) => {
      const summary = item.summary || item.content?.substring(0, 200) || item.title;
      return `${index + 1}. ${item.title}\n   来源: ${item.sourceName}\n   摘要: ${summary}`;
    }).join('\n\n');
  }

  // 构建对话提示
  buildDialoguePrompt(dialogueType, character1, character2, newsSummary, rounds) {
    const typePrompts = {
      interview: `你是一位专业的主持人，正在采访一位IT行业专家。请基于以下新闻内容，进行${rounds}轮深入的访谈对话。主持人应该引导话题，专家应该提供专业的分析和见解。`,
      ceo_interview: `你是一位资深记者，正在采访一位知名企业的CEO。请基于以下新闻内容，进行${rounds}轮专业的采访对话。记者应该提出尖锐的问题，CEO应该给出战略性的回答。`,
      commentary: `你是两位资深的新闻评论员，正在分析最新的新闻热点。请基于以下新闻内容，进行${rounds}轮深入的评论对话。两位评论员应该从不同角度分析问题，提供独到的见解。`,
      chat: `你是两位朋友，正在讨论最新的新闻。请基于以下新闻内容，进行${rounds}轮轻松的聊天对话。对话应该自然、有趣，体现朋友间的交流风格。`
    };

    const basePrompt = typePrompts[dialogueType] || typePrompts.interview;

    return `${basePrompt}

新闻内容：
${newsSummary}

请生成${rounds}轮对话，格式如下：
${character1}: [对话内容]
${character2}: [对话内容]
${character1}: [对话内容]
...

要求：
1. 对话要自然流畅，符合角色设定
2. 内容要基于提供的新闻，但不要直接引用
3. 每轮对话要有深度，体现专业性和见解
4. 对话要有逻辑性，前后呼应
5. 总字数控制在2000字以内`;
  }

  // 调用LLM API
  async callLLMAPI(prompt) {
    try {
      // 获取代理配置
      const proxyConfig = await this.getProxyConfig();
      
      const axiosConfig = {
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json'
        },
        timeout: 60000
      };

      // 如果启用了代理，添加代理配置
      if (proxyConfig) {
        axiosConfig.proxy = proxyConfig;
        logger.info(`使用代理调用LLM API: ${this.apiUrl}, 代理: ${proxyConfig.host}:${proxyConfig.port}`);
      }

      const response = await axios.post(this.apiUrl, {
        model: this.model,
        messages: [
          {
            role: 'system',
            content: '你是一个专业的对话生成助手，能够根据新闻内容生成高质量的对话。'
          },
          {
            role: 'user',
            content: prompt
          }
        ],
        max_tokens: 3000,
        temperature: 0.7
      }, axiosConfig);

      return response.data.choices[0].message.content;
    } catch (error) {
      logger.error('LLM API调用失败:', error);
      throw new Error('LLM服务调用失败');
    }
  }

  // 解析对话响应
  parseDialogueResponse(response, character1, character2) {
    try {
      // 简单的对话解析，按角色分割
      const lines = response.split('\n').filter(line => line.trim());
      const dialogue = [];
      
      for (const line of lines) {
        if (line.startsWith(`${character1}:`)) {
          dialogue.push({
            speaker: character1,
            content: line.substring(character1.length + 1).trim()
          });
        } else if (line.startsWith(`${character2}:`)) {
          dialogue.push({
            speaker: character2,
            content: line.substring(character2.length + 1).trim()
          });
        }
      }
      
      return dialogue;
    } catch (error) {
      logger.error('解析对话响应失败:', error);
      // 如果解析失败，返回原始文本
      return [{
        speaker: '系统',
        content: response
      }];
    }
  }

  // 获取对话类型名称
  getDialogueTypeName(type) {
    const typeNames = {
      interview: '访谈对话',
      ceo_interview: 'CEO采访',
      commentary: '评论对话',
      chat: '聊天对话'
    };
    return typeNames[type] || '对话';
  }

  // 测试LLM连接
  async testConnection() {
    try {
      await this.initialize();
      
      // 获取代理配置
      const proxyConfig = await this.getProxyConfig();
      
      const axiosConfig = {
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json'
        },
        timeout: 10000
      };

      // 如果启用了代理，添加代理配置
      if (proxyConfig) {
        axiosConfig.proxy = proxyConfig;
        logger.info(`使用代理测试LLM连接: ${this.apiUrl}, 代理: ${proxyConfig.host}:${proxyConfig.port}`);
      }

      const response = await axios.post(this.apiUrl, {
        model: this.model,
        messages: [
          {
            role: 'user',
            content: '你好，请回复"连接成功"'
          }
        ],
        max_tokens: 50
      }, axiosConfig);

      return {
        success: true,
        data: {
          responseTime: 1000, // 模拟响应时间
          message: 'LLM连接测试成功',
          response: response.data.choices[0].message.content
        }
      };
    } catch (error) {
      logger.error('LLM连接测试失败:', error);
      return {
        success: false,
        message: 'LLM连接测试失败',
        data: {
          error: error.message
        }
      };
    }
  }
}

module.exports = new LLMService(); 

File: backend-go/internal/services/llm_service.go
package services

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/raciel/rss-llm-tts/internal/config"
	"github.com/raciel/rss-llm-tts/internal/models"
	"github.com/raciel/rss-llm-tts/internal/utils"
	"github.com/rs/zerolog/log"
	"gorm.io/gorm"
)

const (
	llmRequestTimeout = 120 * time.Second
	llmMaxRetries     = 3
	llmMaxContentSize = 1200
)

type LLMService struct {
	db      *gorm.DB
	client  *http.Client
	config  *config.Config
	configs map[string]string
}

type DialogueParams struct {
	DialogueType string
	Character1   string
	Character2   string
	Rounds       int
	NewsContent  []NewsContent
}

type NewsContent struct {
	Title       string
	Summary     string
	Content     string
	Source      string
	PublishedAt time.Time
	Author      string
}

type DialogueResult struct {
	Rounds []Round `json:"rounds"`
}

type Round struct {
	Speaker string `json:"speaker"`
	Text    string `json:"text"`
}

type chatCompletionResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
		FinishReason string `json:"finish_reason"`
	} `json:"choices"`
	Error *struct {
		Message string `json:"message"`
		Type    string `json:"type"`
	} `json:"error,omitempty"`
}

func NewLLMService(db *gorm.DB, cfg *config.Config, configs map[string]string) *LLMService {
	client := &http.Client{Timeout: llmRequestTimeout}
	return &LLMService{
		db:      db,
		client:  client,
		config:  cfg,
		configs: configs,
	}
}

func (s *LLMService) GenerateDialogue(params DialogueParams) (*DialogueResult, error) {
	prompt := s.buildPrompt(params)

	result, err := s.callLLMAPI(prompt)
	if err != nil {
		log.Error().Err(err).Msg("LLM request failed")
		return nil, err
	}

	jsonStr := extractJSONObject(result)
	if jsonStr == "" {
		return nil, fmt.Errorf("llm response did not contain a valid JSON object")
	}

	var dialogue DialogueResult
	if err := json.Unmarshal([]byte(jsonStr), &dialogue); err != nil {
		log.Warn().
			Err(err).
			Str("response", truncateForLog(result, 600)).
			Msg("Failed to parse LLM JSON response")
		return nil, fmt.Errorf("failed to parse llm json response: %w", err)
	}

	if err := validateDialogueResult(&dialogue, params); err != nil {
		return nil, err
	}

	if len(dialogue.Rounds) != params.Rounds {
		log.Warn().
			Int("expected", params.Rounds).
			Int("actual", len(dialogue.Rounds)).
			Msg("LLM returned unexpected round count; normalizing result")
		dialogue.Rounds = s.adjustRounds(dialogue.Rounds, params.Rounds, params.Character1, params.Character2)
	}

	return &dialogue, nil
}

func (s *LLMService) buildPrompt(params DialogueParams) string {
	var newsDetails strings.Builder
	for i, news := range params.NewsContent {
		newsDetails.WriteString(fmt.Sprintf("Article %d\n", i+1))
		newsDetails.WriteString(fmt.Sprintf("Title: %s\n", news.Title))
		if news.Summary != "" {
			newsDetails.WriteString(fmt.Sprintf("Summary: %s\n", news.Summary))
		}
		if news.Source != "" {
			newsDetails.WriteString(fmt.Sprintf("Source: %s\n", news.Source))
		}
		if !news.PublishedAt.IsZero() {
			newsDetails.WriteString(fmt.Sprintf("Published At: %s\n", news.PublishedAt.Format("2006-01-02 15:04:05")))
		}
		if news.Author != "" {
			newsDetails.WriteString(fmt.Sprintf("Author: %s\n", news.Author))
		}
		if news.Content != "" {
			newsDetails.WriteString(fmt.Sprintf("Details: %s\n", truncateForLog(news.Content, llmMaxContentSize)))
		}
		newsDetails.WriteString("\n")
	}

	typeDetails := s.getDialogueTypeDetails(params.DialogueType)

	return fmt.Sprintf(`You are an expert dialogue writer.

Create a high-quality %s conversation in Simplified Chinese.

Requirements:
1. Return valid JSON only, with no markdown fences and no extra commentary.
2. The JSON schema must be exactly: {"rounds":[{"speaker":"string","text":"string"}]}.
3. Produce exactly %d rounds.
4. Speakers must alternate naturally between "%s" and "%s".
5. Each round should be substantial, specific, and grounded in the provided articles.
6. Keep the conversation coherent, insightful, and conversational.
7. Use Simplified Chinese for the dialogue text, except for standard product, company, or technical names when needed.

Dialogue style:
- Type: %s
- Description: %s
- Tone: %s

Articles:
%s`, typeDetails.Name, params.Rounds, params.Character1, params.Character2, typeDetails.Name, typeDetails.Description, typeDetails.Style, newsDetails.String())
}

func (s *LLMService) loadRuntimeConfigs() map[string]string {
	result := map[string]string{
		"llm_api_url":        s.config.LLM.APIURL,
		"llm_api_key":        s.config.LLM.APIKey,
		"llm_model":          s.config.LLM.Model,
		"http_proxy_enabled": strconv.FormatBool(s.config.Proxy.Enabled),
		"http_proxy_url":     s.config.Proxy.URL,
	}

	for key, value := range s.configs {
		if strings.TrimSpace(value) != "" {
			result[key] = value
		}
	}

	var dbConfigs []models.Config
	if err := s.db.Find(&dbConfigs).Error; err != nil {
		log.Warn().Err(err).Msg("Failed to reload configs from database; using cached/runtime values")
		return result
	}

	for _, cfg := range dbConfigs {
		value := cfg.Value
		if cfg.IsEncrypted && value != "" {
			decrypted, err := utils.Decrypt(value)
			if err != nil {
				log.Warn().Err(err).Str("key", cfg.Key).Msg("Failed to decrypt config value; using stored value")
			} else {
				value = decrypted
			}
		}
		if strings.TrimSpace(value) != "" {
			result[cfg.Key] = value
		}
	}

	s.configs = result
	return result
}

func (s *LLMService) callLLMAPI(prompt string) (string, error) {
	configs := s.loadRuntimeConfigs()

	apiURL := firstNonEmpty(configs["llm_api_url"], s.config.LLM.APIURL)
	apiKey := firstNonEmpty(configs["llm_api_key"], s.config.LLM.APIKey)
	model := firstNonEmpty(configs["llm_model"], s.config.LLM.Model)

	if apiURL == "" {
		return "", fmt.Errorf("llm api url is not configured")
	}
	if apiKey == "" {
		return "", fmt.Errorf("llm api key is not configured")
	}
	if model == "" {
		return "", fmt.Errorf("llm model is not configured")
	}

	reqBody := map[string]interface{}{
		"model": model,
		"messages": []map[string]string{
			{
				"role":    "system",
				"content": "You write polished Chinese dialogue and must respond with a JSON object only.",
			},
			{
				"role":    "user",
				"content": prompt,
			},
		},
		"temperature": 0.3,
		"max_tokens":  4000,
		"top_p":       0.9,
		"response_format": map[string]string{
			"type": "json_object",
		},
	}

	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return "", err
	}

	proxyURL := resolveProxyURL(configs["http_proxy_enabled"], configs["http_proxy_url"])
	client := s.newHTTPClient(proxyURL)

	var lastErr error
	for attempt := 1; attempt <= llmMaxRetries; attempt++ {
		ctx, cancel := context.WithTimeout(context.Background(), llmRequestTimeout)
		req, err := http.NewRequestWithContext(ctx, http.MethodPost, apiURL, bytes.NewReader(bodyBytes))
		if err != nil {
			cancel()
			return "", err
		}

		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Authorization", "Bearer "+apiKey)

		resp, err := client.Do(req)
		if err != nil {
			cancel()
			lastErr = fmt.Errorf("llm request attempt %d failed: %w", attempt, err)
			log.Warn().Err(lastErr).Int("attempt", attempt).Msg("LLM request error")
			if attempt < llmMaxRetries {
				time.Sleep(time.Duration(attempt) * time.Second)
			}
			continue
		}

		responseBody, readErr := io.ReadAll(resp.Body)
		resp.Body.Close()
		cancel()
		if readErr != nil {
			lastErr = fmt.Errorf("failed to read llm response body: %w", readErr)
			if attempt < llmMaxRetries {
				time.Sleep(time.Duration(attempt) * time.Second)
			}
			continue
		}

		if resp.StatusCode >= http.StatusBadRequest {
			lastErr = fmt.Errorf("llm api status %d: %s", resp.StatusCode, truncateForLog(string(responseBody), 800))
			log.Warn().Err(lastErr).Int("attempt", attempt).Msg("LLM returned non-success status")
			if attempt < llmMaxRetries {
				time.Sleep(time.Duration(attempt) * time.Second)
			}
			continue
		}

		var parsed chatCompletionResponse
		if err := json.Unmarshal(responseBody, &parsed); err != nil {
			lastErr = fmt.Errorf("failed to decode llm api response: %w", err)
			log.Warn().
				Err(lastErr).
				Int("attempt", attempt).
				Str("body", truncateForLog(string(responseBody), 800)).
				Msg("Invalid LLM API response JSON")
			if attempt < llmMaxRetries {
				time.Sleep(time.Duration(attempt) * time.Second)
			}
			continue
		}

		if parsed.Error != nil && parsed.Error.Message != "" {
			lastErr = fmt.Errorf("llm api error: %s", parsed.Error.Message)
			if attempt < llmMaxRetries {
				time.Sleep(time.Duration(attempt) * time.Second)
			}
			continue
		}

		if len(parsed.Choices) == 0 {
			lastErr = fmt.Errorf("llm api returned no choices")
			if attempt < llmMaxRetries {
				time.Sleep(time.Duration(attempt) * time.Second)
			}
			continue
		}

		content := strings.TrimSpace(parsed.Choices[0].Message.Content)
		if content == "" {
			lastErr = fmt.Errorf("llm api returned empty message content")
			if attempt < llmMaxRetries {
				time.Sleep(time.Duration(attempt) * time.Second)
			}
			continue
		}

		return content, nil
	}

	if lastErr == nil {
		lastErr = fmt.Errorf("llm request failed for an unknown reason")
	}
	return "", lastErr
}

func (s *LLMService) newHTTPClient(proxyURL *url.URL) *http.Client {
	transport := &http.Transport{}
	if proxyURL != nil {
		transport.Proxy = http.ProxyURL(proxyURL)
	}

	return &http.Client{
		Timeout:   llmRequestTimeout,
		Transport: transport,
	}
}

func (s *LLMService) getDialogueTypeDetails(dialogueType string) DialogueTypeDetails {
	detailsMap := map[string]DialogueTypeDetails{
		"interview": {
			Name:        "interview",
			Description: "A focused interviewer and guest discussion with clear questions, analysis, and takeaways.",
			Style:       "professional, analytical, and engaging",
		},
		"ceo_interview": {
			Name:        "ceo interview",
			Description: "A business-oriented conversation about strategy, execution, leadership, and market direction.",
			Style:       "strategic, executive, and insightful",
		},
		"commentary": {
			Name:        "commentary",
			Description: "A current-events conversation that compares viewpoints and explains implications.",
			Style:       "sharp, balanced, and thoughtful",
		},
		"chat": {
			Name:        "chat",
			Description: "A relaxed but substantive conversation that still delivers clear ideas and depth.",
			Style:       "natural, conversational, and clear",
		},
	}

	if details, ok := detailsMap[dialogueType]; ok {
		return details
	}

	return detailsMap["interview"]
}

type DialogueTypeDetails struct {
	Name        string
	Description string
	Style       string
}

func (s *LLMService) generateMockDialogue(params DialogueParams) *DialogueResult {
	rounds := make([]Round, params.Rounds)
	typeDetails := s.getDialogueTypeDetails(params.DialogueType)

	for i := 0; i < params.Rounds; i++ {
		speaker := params.Character1
		if i%2 == 1 {
			speaker = params.Character2
		}

		newsIndex := 0
		if len(params.NewsContent) > 0 {
			newsIndex = i % len(params.NewsContent)
		}

		title := "the topic"
		if len(params.NewsContent) > 0 && params.NewsContent[newsIndex].Title != "" {
			title = params.NewsContent[newsIndex].Title
		}

		text := fmt.Sprintf("This is a placeholder %s round about %s.", typeDetails.Name, title)
		rounds[i] = Round{Speaker: speaker, Text: text}
	}

	return &DialogueResult{Rounds: rounds}
}

func (s *LLMService) adjustRounds(rounds []Round, target int, char1, char2 string) []Round {
	if target <= 0 {
		return []Round{}
	}

	if len(rounds) > target {
		return rounds[:target]
	}

	for len(rounds) < target {
		speaker := char1
		if len(rounds)%2 == 1 {
			speaker = char2
		}
		rounds = append(rounds, Round{
			Speaker: speaker,
			Text:    "Please continue the discussion with a concrete point tied to the news.",
		})
	}

	return rounds
}

func extractJSONObject(raw string) string {
	trimmed := strings.TrimSpace(raw)
	start := strings.Index(trimmed, "{")
	end := strings.LastIndex(trimmed, "}")
	if start == -1 || end == -1 || end < start {
		return ""
	}

	return strings.TrimSpace(trimmed[start : end+1])
}

func validateDialogueResult(dialogue *DialogueResult, params DialogueParams) error {
	if dialogue == nil {
		return fmt.Errorf("llm returned an empty dialogue result")
	}
	if len(dialogue.Rounds) == 0 {
		return fmt.Errorf("llm returned zero dialogue rounds")
	}

	for i := range dialogue.Rounds {
		dialogue.Rounds[i].Speaker = strings.TrimSpace(dialogue.Rounds[i].Speaker)
		dialogue.Rounds[i].Text = strings.TrimSpace(dialogue.Rounds[i].Text)

		if dialogue.Rounds[i].Speaker == "" {
			if i%2 == 0 {
				dialogue.Rounds[i].Speaker = params.Character1
			} else {
				dialogue.Rounds[i].Speaker = params.Character2
			}
		}

		if dialogue.Rounds[i].Text == "" {
			return fmt.Errorf("llm returned an empty text for round %d", i+1)
		}
	}

	return nil
}

func truncateForLog(value string, limit int) string {
	if limit <= 0 || len(value) <= limit {
		return value
	}
	return value[:limit] + "..."
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return strings.TrimSpace(value)
		}
	}
	return ""
}

func resolveProxyURL(enabledValue, proxyValue string) *url.URL {
	if !parseBool(enabledValue) || strings.TrimSpace(proxyValue) == "" {
		return nil
	}

	proxyURL, err := url.Parse(strings.TrimSpace(proxyValue))
	if err != nil {
		log.Warn().Err(err).Str("proxy", proxyValue).Msg("Invalid proxy URL; ignoring proxy")
		return nil
	}

	return proxyURL
}

func parseBool(value string) bool {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}
